src/searcher/core.rs

```diff
@@ -0,0 +1,258 @@
+        if self.config.before_context == 0 {
+            return Ok(true);
+        }
+        let range = Range::new(self.last_line_visited, upto);
+        if range.is_empty() {
+            return Ok(true);
+        }
+        let before_context_start = range.start()
+            + lines::preceding(
+                &buf[range],
+                self.config.line_term.as_byte(),
+                self.config.before_context - 1,
+            );
+
+        let range = Range::new(before_context_start, range.end());
+        let mut stepper =
+            LineStep::new(self.config.line_term.as_byte(), range.start(), range.end());
+        while let Some(line) = stepper.next_match(buf) {
+            if !self.sink_break_context(line.start())? {
+                return Ok(false);
+            }
+            if !self.sink_before_context(buf, &line)? {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn after_context_by_line(&mut self, buf: &[u8], upto: usize) -> Result<bool, S::Error> {
+        if self.after_context_left == 0 {
+            return Ok(true);
+        }
+        let range = Range::new(self.last_line_visited, upto);
+        let mut stepper =
+            LineStep::new(self.config.line_term.as_byte(), range.start(), range.end());
+        while let Some(line) = stepper.next_match(buf) {
+            if !self.sink_after_context(buf, &line)? {
+                return Ok(false);
+            }
+            if self.after_context_left == 0 {
+                break;
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn other_context_by_line(&mut self, buf: &[u8], upto: usize) -> Result<bool, S::Error> {
+        let range = Range::new(self.last_line_visited, upto);
+        let mut stepper =
+            LineStep::new(self.config.line_term.as_byte(), range.start(), range.end());
+        while let Some(line) = stepper.next_match(buf) {
+            if !self.sink_other_context(buf, &line)? {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    #[inline(always)]
+    fn sink_matched(
+        &mut self,
+        buf: &[u8],
+        range: &Range,
+        exact_matches: &[Range],
+    ) -> Result<bool, S::Error> {
+        if !self.sink_break_context(range.start())? {
+            return Ok(false);
+        }
+        self.count_lines(buf, range.start());
+        let offset = self.absolute_byte_offset + range.start() as u64;
+        let linebuf = &buf[*range];
+        let keepgoing = self.sink.matched(
+            self.searcher,
+            &SinkMatch {
+                line_term: self.config.line_term,
+                bytes: linebuf,
+                absolute_byte_offset: offset,
+                line_number: self.line_number,
+                buffer: buf,
+                bytes_range_in_buffer: range.start()..range.end(),
+                exact_matches,
```

Comment:
Here's where we're passing the `exact_matches` that we've been keeping track of to the `Sink` API